pkg/synth/acl.go

```diff
@@ -15,153 +15,177 @@ import (
+			request = &ir.Packet{Src: srcSubnet.Addrs, Dst: dstCidr, Protocol: protocol, Explanation: reason.String()}
+			addRuleToACL(result, ir.AllowSend(*request), srcSubnet, internal, singleACL)
 			if inverseProtocol := protocol.InverseDirection(); inverseProtocol != nil {
-				response = &ir.Packet{Src: dstCidr, Dst: srcSubnetCidr, Protocol: inverseProtocol, Explanation: reason.response().String()}
-				connection = append(connection, ir.AllowReceive(*response))
+				response = &ir.Packet{Src: dstCidr, Dst: srcSubnet.Addrs, Protocol: inverseProtocol, Explanation: reason.response().String()}
+				addRuleToACL(result, ir.AllowReceive(*response), srcSubnet, internal, singleACL)
 			}
 		}
 	}
 
 	if internalDst {
-		for _, dstSubnetCidr := range dstSubnetList {
-			if srcCidr == dstSubnetCidr {
+		for _, dstSubnet := range dstSubnets {
+			if dstSubnet.Addrs.Equal(srcCidr) { // dstSubnet and srcCidr are the same subnet
 				continue
 			}
-			request = &ir.Packet{Src: srcCidr, Dst: dstSubnetCidr, Protocol: protocol, Explanation: reason.String()}
-			connection = append(connection, ir.AllowReceive(*request))
+			request = &ir.Packet{Src: srcCidr, Dst: dstSubnet.Addrs, Protocol: protocol, Explanation: reason.String()}
+			addRuleToACL(result, ir.AllowReceive(*request), dstSubnet, internal, singleACL)
 			if inverseProtocol := protocol.InverseDirection(); inverseProtocol != nil {
-				response = &ir.Packet{Src: dstSubnetCidr, Dst: srcCidr, Protocol: inverseProtocol, Explanation: reason.response().String()}
-				connection = append(connection, ir.AllowSend(*response))
+				response = &ir.Packet{Src: dstSubnet.Addrs, Dst: srcCidr, Protocol: inverseProtocol, Explanation: reason.response().String()}
+				addRuleToACL(result, ir.AllowSend(*response), dstSubnet, internal, singleACL)
 			}
 		}
 	}
-
-	return connection
 }
 
-func generateACLCollectionForBlockedSubnets(s *ir.Spec, aclSelector func(target *ipblock.IPBlock) string) *ir.ACLCollection {
+// generate a nACL collection, which include nACL(s) rules that block all traffic to/from the blocked subnets
+func generateACLCollectionForBlockedSubnets(s *ir.Spec, single bool) *ir.ACLCollection {
 	blockedSubnets := s.ComputeBlockedSubnets()
 	result := ir.NewACLCollection()
 	for _, subnet := range blockedSubnets {
+		acl := result.LookupOrCreate(aclSelector(subnet, single))
 		cidr := s.Defs.Subnets[subnet].Address()
-		acl := result.LookupOrCreate(aclSelector(cidr))
 		acl.AppendInternal(ir.DenyAllReceive(subnet, cidr))
 		acl.AppendInternal(ir.DenyAllSend(subnet, cidr))
 	}
-
 	return result
 }
 
-func resourceRelevantToACL(e ir.ResourceType) bool {
-	return e == ir.ResourceTypeSubnet || e == ir.ResourceTypeCidr || e == ir.ResourceTypeNIF
+// Change ir.Resource to a slice of ir.ConnResource (each ir.ConnResource is either a pair of a of a subnet with its CIDR or an
+// external with its IPAddrs), to make it convenient to go through all pairs of subnets in allowDirectedConnection function.
```

Comment:
Also explain what is the `addrs` parameter and why/when it is changed in the returned values.